### PR TITLE
feat: persist rate limiting

### PR DESCRIPTION
**Pull request by iamacook**
This replaces `@nestjs/throttler` with a custom rate limiter, persisted with Redis:

- Removes `@nestjs/throttler`.
- Adds `redis`.
- Adds global `CacheModule`, exposing `set` and `get` methods.
- Updates `RelayLimitService` to use caching.
- Associated tests and CI adjustments.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ on:
   push:
   pull_request:
   release:
-    types: [ released ]
+    types: [released]
```
Comment (iamacook) [inline .github/workflows/ci.yml:7]: Side note: it's probably wise we add a Prettier config to this repo at some stage.

File: src/routes/relay/services/relay-limit.service.ts [hunk 2]
```diff
@@ -26,47 +30,65 @@ export class RelayLimitService {
...
+      addresses.map((address) => this.getCachedAttempts(chainId, address)),
+    );
+
     await Promise.all(
-      addresses.map((address) => {
+      addresses.map((address, i) => {
         const key = this.generateKey(chainId, address);
-        return this.throttlerStorageService.increment(key, this.ttl);
+        const incremented = attempts[i] + 1;
+        return this.cacheService.set(key, incremented.toString(), this.ttl);
```
Comment (fmrsabino) [inline src/routes/relay/services/relay-limit.service.ts:91]: I think we can merge the two steps above. What do you think about the following?

```typescript
await Promise.all(
      addresses.map(async (address) => {
        const key: string = this.generateKey(chainId, address);
        const currentAttempts: number = await this.getCachedAttempts(
          chainId,
          address,
        );
        const incremented: number = currentAttempts + 1;
        return this.cacheService.set(key, incremented.toString(), this.ttl);
      }),
    );
```

Another thing to consider is in the same way that we have a `getCachedAttempts` returning a number we can probably have a `setCachedAttempts` which takes a number. With that, we "normalize" the read/write access to the cache and hide the `string` representation behind these two functions.

One thing to consider though is that the attempts might get partially updated – if one fails this promise is rejected. What do you think about using `Promise.allSettled()` instead?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (fmrsabino) [reply .github/workflows/ci.yml:7]: Good point. I didn't know that it supported YAML. Did you apply it here for this PR?

**Review by fmrsabino — CHANGES_REQUESTED** [review]
Nice! Have just some small comments!

File: .github/workflows/ci.yml (hunk 1 above)
Comment (iamacook) [reply .github/workflows/ci.yml:7]: My IDE did it automatically.

File: src/routes/relay/services/relay-limit.service.ts (hunk 2 above)
Comment (iamacook) [reply src/routes/relay/services/relay-limit.service.ts]: Great ideas! I've combined them all as follows:

- There is now a `setCachedAttempts` function.
- `increment` leverages this, combining both the getting and setting inside one `map`.
- All of which is wrapped in a `Promise.allSettled`.

File: src/datasources/cache/__tests__/test.cache.module.ts [hunk 3]
```diff
@@ -1,9 +1,9 @@
 import { Global, Module } from '@nestjs/common';
 
 import { CacheService } from '../cache.service.interface';
-import { MockCacheService } from './mock.cache.service';
+import { FakeCacheService } from './fake.cache.service';
 
-const mockCacheService = new MockCacheService();
+const mockCacheService = new FakeCacheService();
```
Comment (fmrsabino) [inline src/datasources/cache/__tests__/test.cache.module.ts:6]: nitpick: `fakeCacheService` 😄

**Review by fmrsabino — APPROVED** [review]
Nice! 🚀 🚀 🚀

**Review by fmrsabino — APPROVED** [review]

**Pull request merged**